### ISSUE TITLE
[C++ Verification] Can't construct rvalue reference to array type during dereference

### ISSUE THREAD

**Issue by ChenfengWei0** [esbmc/esbmc#1934]
```c++
class Base
{
public:
  int ss[128];
  Base() 
  {
    ss;
  }
};

int main()
{
  Base x;
}
```

I got 
```sh
ESBMC version 7.6.1 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing 1.cpp
1.cpp:13:5: warning: expression result unused [-Wunused-value]
    ss;
    ^~
Converting
Generating GOTO Program
GOTO program creation time: 0.455s
GOTO program processing time: 0.001s
Starting Bounded Model Checking
ERROR: Can't construct rvalue reference to array type during dereference
(It isn't allowed by C anyway)

[1]    11807 IOT instruction  ./esbmc 1.cpp
```

If I comment the abort statements in `dereference.cpp`
```c++
  else if (is_array_type(type))
  {
     // log_error(
     //  "Can't construct rvalue reference to array type during dereference\n"
     // "(It isn't allowed by C anyway)\n");
    // abort();
    ;
  }
```
then everything works fine, including the case shown below

```c++
#include <assert.h>
class Base
{
public:
  int ss[128];
  Base() : ss{0} 
  {
  }
};

int main()
{
  Base x;
  assert(x.ss[1] == 0); // hold
  assert(x.ss[0] == 1); // fail
}
```

@XLiZHI , do you have any ideas on this? Since I need this feature for Solidity, would it be okay if I just comment the abort?

**Comment by lucasccordeiro** [comment]
I think @fbrausse can also contribute to this discussion.

**Comment by fbrausse** [comment]
Nice, thanks @ChenfengWei0. We've been looking for examples where this error happens! I'll take a look.

**Comment by fbrausse** [comment]
> Since I need this feature for Solidity, would it be okay if I just comment the abort?

Commenting the abort() would mean that the type of each such dereference is being treated as a scalar (e.g. `int`), while it actually is an array. This is a tricky piece of code and we've fixed many similar occurrences of this error message in the past. Let's investigate it first.

**Comment by XLiZHI** [comment]
https://github.com/esbmc/esbmc/blob/de10dee30288a464dbbfa2dfa6575ff3522130e6/src/clang-c-frontend/clang_c_adjust_expr.cpp#L254-L266

I have encountered this problem before and it should be due to this line of code: `ss;`

~~The adjust member code here unconditionally dereferences the array type, which is what caused the error~~.

```cpp
    0: dereference
        * type: symbol
            * identifier: tag-Base
        * operands: 
          0: symbol
              * type: pointer
                  * subtype: symbol
                      * identifier: tag-Base
              * name: this
              * identifier: c:@S@Base@F@Base#this
              * #location: 
                * file: main6.cpp
                * line: 7
                * function: Base
                * column: 5
              * #lvalue: 1
```
If I'm right, the goto function here ~~should be~~: `this->ss[0];` rather than `this->ss;`

I dug a little deeper and we need to figure out how to assign values to the array type: `this->ss = {0,0,0,0...};` Converting to  `this->ss[0];`doesn't work, I also tried CBMC and it didn't seem to have this error?

_Edit:_ I tried several solutions, should we consider whether dereferencing of array types is correct in C++?

**Comment by fbrausse** [comment]
> _Edit:_ I tried several solutions, should we consider whether dereferencing of array types is correct in C++?

We can consider it, but the point is: there is no implementation in the dereference module that handles this case right now. It would need to be implemented. I don't fully understand why it hasn't been done in the past, but I can think of these (somewhat overlapping) reasons:
1. In C, which was originally the only target, it's illegal¹ to assign to/from (syntactical) arrays. There's some lore around why: https://stackoverflow.com/questions/35597019/cs-aversion-to-arrays/35598701#35598701 The short version is: backwards compatibility with B and BCPL. We still pretend to have it today. ;)

   ¹ This is really hard to make precise because the C standard semantically treats all objects as arrays, of size 1, if necessary, when talking about pointers.
3. LValues are never arrays in C, because they decay to pointers. This is of course different from C++ which has the concept of "references" - as you well know. ;)

    From C90:
    > 6.2.2.1 "A modifiable lvalue is an lvalue that does not have array type, does not have an incomplete type, does not have a const-qualified type. and if it is a structure or union. does not have any member (including. recursively, any member of all contained structures or unions) with a const-qualified type."
4. There are arrays without bounds (flexible array members in structs) or with runtime bounds (variable length arrays). Copying those in a C program requires runtime support, which wasn't deemed to be feasible at the time. For us, those mean extra checks: are the source and target actually of compatible size?
5. Because of the points above, there are no semantics to be inferred from the standard, or from GCC, regarding aliasing, alignment, or out-of-bounds accesses. My best suggestion right now would be to treat it as a memcpy().

However, as @mikhailramalho did point out to me some time ago, ESBMC's irep and irep2 isn't modelling C, it's rather supposed to be very close to SMT or any other formal language that knows about equalities and data types *without implicit casts* (because they are the root of all evil). So the reasoning in the first two points above isn't really cutting it. But we still need semantics: after all, the dereference module is supposed to check whether all dereferenced accesses are correct.

My conclusion: Yes, we could actually allow arrays to be the target of dereferences, however, this needs an implementation, and this implementation needs to be sound (see items 3. and 4. above).

**Comment by intrigus-lgtm** [comment]
```cpp
#include <assert.h>
struct Base {
    int ss[1] = {0};
};

int main() {
    // Base x;
    // assert(x.ss[0] == 0);
    int aa[1] = {0};
    assert(aa[0] == 0);
}
```

**Interestingly, esbmc allows assigning arrays outside of structs (`this->ss` vs`aa`).**

<details>
<summary>IR (fails)</summary>

```cpp
Base (c:@S@Base@F@Base#):
        // 20 no location
        THROW_DECL (noexcept)
        // 21 
        this->ss={ 0 };
        // 22 file lol.cpp line 2 column 8 function Base
        END_FUNCTION // Base
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

main (c:@F@main#):
        // 23 file lol.cpp line 7 column 5 function main
        Base x;
        // 24 file lol.cpp line 7 column 10 function main
        Base tmp$1;
        // 25 file lol.cpp line 7 column 10 function main
        tmp$1=NONDET(struct);
        // 26 
        FUNCTION_CALL:  Base(&tmp$1)
        // 27 file lol.cpp line 7 column 5 function main
        x=tmp$1;
        // 28 file lol.cpp line 8 column 5 function main
        ASSERT x.ss[0] == 0 // assertion x.ss[0] == 0
        // 29 file lol.cpp line 8 column 5 function main
        IF !(x.ss[0] == 0) THEN GOTO 1
        // 30 file lol.cpp line 8 column 5 function main
        0;

        // 31 file lol.cpp line 9 column 5 function main
     1: signed int [1] aa;
        // 32 file lol.cpp line 9 column 5 function main
        aa={ 0 };
        // 33 file lol.cpp line 10 column 5 function main
        ASSERT aa[0] == 0 // assertion aa[0] == 0
```
</details>

**Comment by intrigus-lgtm** [comment]
Another question would be: what does the old frontend do?
(I'm currently on a branch where I can't build the old frontend, so I can't test myself)

**Comment by fbrausse** [comment]
> Interestingly, esbmc allows assigning arrays outside of structs (`this->ss` vs`aa`).

Initialization is fine, assignments aren't. The error here happens when a dereference expression is array-typed.

> Another question would be: what does the old frontend do?

The same error. These are the relevant GOTO functions for the original program with the old frontend:
```
Base (Base::Base(this)):
        // 1 file gh-1934.cc line 7
        this->ss;

        // 2 no location
        END_FUNCTION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

main (main()):
        // 3 file gh-1934.cc line 13
        Base x;
        // 4 file gh-1934.cc line 13
        x=NONDET(struct);
        // 5 file gh-1934.cc line 13
        FUNCTION_CALL:  Base(&x)
        // 6 no location
        FUNCTION_CALL:  ~Base(&x)
        // 7 no location
        dead x;
        // 8 no location
        RETURN: NONDET(signed int)
        // 9 no location
        END_FUNCTION
```

This even fails for C (old and new frontends)
```c
struct Base
{
  int ss[128];
};

int main()
{
  struct Base x, *y = &x;
  y->ss; // works for &y->ss[0], which the implicit array-to-pointer decay (only in C) should do here
}
```

though here the last statement in main() Clang treats as decayed pointer
```
    `-ImplicitCastExpr 0x558b51f00bf0 <line:9:3, col:6> 'int *' <ArrayToPointerDecay>
      `-MemberExpr 0x558b51f00b90 <col:3, col:6> 'int [128]' lvalue ->ss 0x558b51f00828
        `-ImplicitCastExpr 0x558b51f00b78 <col:3> 'struct Base *' <LValueToRValue>
          `-DeclRefExpr 0x558b51f00b58 <col:3> 'struct Base *' lvalue Var 0x558b51f00a88 'y' 'struct Base *'
```
while interpreting it as C++ it doesn't:
```
    `-MemberExpr 0x55906ee2ad30 <line:9:3, col:6> 'int [128]' lvalue ->ss 0x55906ee23d78
      `-ImplicitCastExpr 0x55906ee2ad18 <col:3> 'struct Base *' <LValueToRValue>
        `-DeclRefExpr 0x55906ee2acf8 <col:3> 'struct Base *' lvalue Var 0x55906ee2ac28 'y' 'struct Base *'
```

**Comment by fbrausse** [comment]
The current build_reference_rec() has been introduced in 4205091afc4e9c5f9660a1b4a23421463a3ea799 as a proposal. From my understanding of the old code (I didn't get it to compile), it would in the case of this issue here first do this:
https://github.com/esbmc/esbmc/blob/4205091afc4e9c5f9660a1b4a23421463a3ea799/pointer-analysis/dereference.cpp#L732 to build a member expression, followed by recursing into build_reference_rec() again, then
https://github.com/esbmc/esbmc/blob/4205091afc4e9c5f9660a1b4a23421463a3ea799/pointer-analysis/dereference.cpp#L751 which would see it's not an array element (type too big for array's subtype) or misaligned, and call
https://github.com/esbmc/esbmc/blob/4205091afc4e9c5f9660a1b4a23421463a3ea799/pointer-analysis/dereference.cpp#L1032-L1033
So basically, emulating memcpy(), but only for arrays of statically known constant size.

**Comment by fbrausse** [comment]
Comparing with CBMC isn't too helpful: they parse it into an 'OTHER expression'
```
        // 3 file gh-1934-1.c line 9 function main
        OTHER code #source_location="" statement="expression"(*main::1::y.ss)
```
which during symex just gets ignored:
```c++
  if(statement==ID_expression)
  {
    // ignore
  }
```

Therefore, it fails to see `(*(struct Base *volatile *)0)->ss` as an error, even though it fails at runtime.

**Comment by fbrausse** [comment]
Hi @ChenfengWei0! The PR #1938 I opened, and which the CI reports as "green", (and which @XLiZHI appearently already explored before) could fix the immediate problem you've been seeing, but it doesn't fix all problems related to C++ class initialization. We're looking into another way to fix those, but maybe that PR could already help you to continue your work?

**Comment by ChenfengWei0** [comment]
> Hi @ChenfengWei0! The PR #1938 I opened, and which the CI reports as "green", (and which @XLiZHI appearently already explored before) could fix the immediate problem you've been seeing, but it doesn't fix all problems related to C++ class initialization. We're looking into another way to fix those, but maybe that PR could already help you to continue your work?

Sure, I will try it out. Many thanks, @fbrausse !
